core/src/main/java/software/amazon/awssdk/auth/profile/ProfileFile.java

```diff
@@ -0,0 +1,348 @@
+                                                                       .build()));
+    }
+
+    private static void addConfigFile(ProfileFile.Aggregator builder) {
+        ProfileFileLocations.configurationFileLocation()
+                            .ifPresent(l -> builder.addFile(ProfileFile.builder()
+                                                                       .content(l)
+                                                                       .type(ProfileFile.Type.CONFIGURATION)
+                                                                       .build()));
+    }
+
+    /**
+     * Convert the raw profile contents into {@link Profile} objects.
+     */
+    private Map<String, Profile> convertRawProfiles(Map<String, Map<String, String>> rawProfiles) {
+        Map<String, Map<String, String>> sortedProfiles = sortProfilesWithParentsFirst(rawProfiles);
+        Map<String, Profile> result = convertToProfilesMap(sortedProfiles);
+
+        return Collections.unmodifiableMap(result);
+    }
+
+    /**
+     * Return a linked hash map that guarantees a profile is always encountered after its parent when iterating over the entry
+     * set. This is useful because a child profile needs access to its parent profile when loading credentials, so we want to
+     * initialize the parent profiles first.
+     */
+    private Map<String, Map<String, String>> sortProfilesWithParentsFirst(Map<String, Map<String, String>> unsortedProfiles) {
+        Map<String, Map<String, String>> sortedProfiles = new LinkedHashMap<>();
+
+        while (!unsortedProfiles.isEmpty()) {
+            String nextProfileToSort = unsortedProfiles.keySet().iterator().next();
+            sortProfileAndParents(unsortedProfiles, sortedProfiles, nextProfileToSort, new HashSet<>());
+        }
+
+        return sortedProfiles;
+    }
+
+    /**
+     * Remove the requested profile and its parents from the unsorted list and put them in the sorted list in an order such
+     * that the parents occur earlier than the children.
+     *
+     * @param unsorted The unsorted collection of profiles so far.
+     * @param sorted The sorted collection of profiles so far.
+     * @param profileName The name of the profile to move to the sorted collection.
+     * @param children Any known children of this profile, which will allow us to ensure there is no circular references in
+     * the configuration file.
+     */
+    private void sortProfileAndParents(Map<String, Map<String, String>> unsorted,
+                                       Map<String, Map<String, String>> sorted,
+                                       String profileName,
+                                       Set<String> children) {
+        Validate.validState(!children.contains(profileName),
+                            "Invalid profile file: Circular relationship detected with profiles %s.", children);
```

Comment:
How far down the chain does this go? if I have a -> b -> c -> d -> a  will that be detected?

Reply:
It's recursive, so our stack size is the limit. It's just checking that children of the 'parent' profiles does not include itself.

We could put in a specific depth if we're worried about someone having 10,000 related profiles in a file and causing a stack overflow (and really slow credential resolution).

Reply:
not too worried about 10,000 credential files.